### PR TITLE
Allow usernames to be 25 chars max with emojis

### PR DESCRIPTION
**Pull request by BendingBender**
- refactor emoji counting text field as a separate component

### CODE REVIEW

File: lib/widgets/text_field_with_max_length.dart [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+class _TextFieldWithMaxLengthState extends State<TextFieldWithMaxLength> {
+  int _currentLength = 0;
+  late final TextEditingController _controller;
+
+  @override
+  void initState() {
+    super.initState();
+
+    _controller = widget.controller ?? TextEditingController();
+    _currentLength = _controller.text.characters.length;
```
Comment (cc-cpo) [inline lib/widgets/text_field_with_max_length.dart]: not sure if the value will be checked against maxLength here?

File: lib/widgets/text_field_with_max_length.dart [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+      ),
+    );
+  }
+}
+
+class _StringCharactersLimitingTextInputFormatter extends TextInputFormatter {
+  final int maxLength;
+
+  _StringCharactersLimitingTextInputFormatter(this.maxLength)
+      : assert(maxLength == -1 || maxLength > 0);
```
Comment (cc-cpo) [inline lib/widgets/text_field_with_max_length.dart:77]: maxLengthcould be an optional

File: lib/widgets/text_field_with_max_length.dart [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+  @override
+  TextEditingValue formatEditUpdate(
```
Comment (cc-cpo) [inline lib/widgets/text_field_with_max_length.dart:80]: how's the component responding to pasted text at the end/in middle position?

File: lib/widgets/text_field_with_max_length.dart (hunk 2 above)
Comment (BendingBender) [reply lib/widgets/text_field_with_max_length.dart:77]: no, not necessary here, this is a private class

File: lib/widgets/text_field_with_max_length.dart (hunk 3 above)
Comment (BendingBender) [reply lib/widgets/text_field_with_max_length.dart:80]: it truncates

File: lib/widgets/text_field_with_max_length.dart (hunk 1 above)
Comment (BendingBender) [reply lib/widgets/text_field_with_max_length.dart]: done

**Pull request merged**